app/models/category.rb

```diff
@@ -0,0 +1,25 @@
+class Category < ApplicationRecord
+  has_and_belongs_to_many :products
+
+  before_validation :remove_white_space_from_name
+
+  validates :name,  presence: true,
+  length: { minimum: 1 },
+  uniqueness: { case_sensitive: false }
+  validate :products_or_error
+
+
+  private
+
+  def products_or_error
+    self.products.each { |product| product_or_error(product) }
+  end
+
+  def product_or_error(new_product)
+    raise ArgumentError.new("Invalid product") if !new_product.is_a?(Product)
+  end
+
+  def remove_white_space_from_name
+    self.name.squish! if !self.name.nil?
```

Comment:
Just a little trick, but you could probably call this method in `after_validation` instead of `before_validation`... your validation step already checks if `name` is going to be nil, so you can just turn this method into `self.name.squish!`